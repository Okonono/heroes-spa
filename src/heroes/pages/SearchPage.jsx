import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../components';

import { getHerosByName } from '../helpers';
import queryString from 'query-string';


export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();
    
    const { q = ''} = queryString.parse (location.search);
    const heroes = getHerosByName(q);

    const showSearch = (q.length === 0 ) // YA REGRESA T O F ? false : true;
    const showError = (q.length > 0) && heroes.length === 0;
    const { buscarTexto, onInputChange } = useForm ({
      buscarTexto: q
    });

    

    const enBusquedaSubmit = (e) => {
        e.preventDefault(); //evitar propagacion del formulario - evitar que haga un refresh
        //if(buscarTexto.trim().length <=1 ) return;
        //console.log( { buscarTexto} );
        navigate( `?q=${ buscarTexto.toLowerCase() }` );

}

  return (
    <>
      <h1>Búsquedas</h1>
      <hr/>

    <div className="row">
          <div className="col-5">
              <h4>Buscar</h4>
              <hr />
              <form onSubmit={ enBusquedaSubmit } aria-label='form'>
                <input 
                  type="text"
                  placeholder="Buscar aquí"
                  className="form-control"
                  name="buscarTexto"
                  autoComplete="off"
                  value={ buscarTexto }
                  onChange={ onInputChange }
                />

                <button className="btn btn-outline-info mt-2">
                    Buscar
                </button>
              </form>
           </div>

           <div className="col-7">
                <h4>Resultados</h4>
                <hr/>

                {/* {
                  ( q=== '')
                  ? <div className="alert alert-primary">Buscar un héroe</div>
                  :  (heroes.length === 0) 
                  && <div className="alert alert-danger">No hay resultados con <b> { q } </b> </div>               
                } */}

                <div className="alert alert-primary animate__animated animate__fadeIn" 
                    style={ { display: showSearch ? '' : 'none' }}
                  >Buscar un héroe
                </div>

                <div aria-label='pikachu' className="alert alert-danger animate__animated animate__fadeIn" 
                    style={ { display: showError ? '' : 'none'  } }
                  >No hay resultados con <b> { q } </b> 
                  
                </div>



                {
                  heroes.map ( hero =>(
                    <HeroCard key={ hero.id } { ...hero }/>
                  ))
                }


          </div>

      </div>
    </>
  )
}
