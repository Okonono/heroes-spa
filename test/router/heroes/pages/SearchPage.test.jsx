import { fireEvent, getByRole, render,screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../../src/heroes/pages/SearchPage';


const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}) );

describe('Pruebas en <SearchPage />', () => { 

    beforeEach( () => jest.clearAllMocks() );
    
    test('debe mostrarse correctamente con valores por defecto', () => { 

        const { container } =render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        );
        expect( container ).toMatchSnapshot();
    });


    test('debe mostrar a Batman y el input con el valor del queryString', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage/>
            </MemoryRouter>
        );
       
        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');
        
        const imagen = screen.getByRole('img');
        expect( imagen.src ).toContain('/assets/heroes/dc-batman.jpg');

        const  pikachuTest = screen.getByLabelText('pikachu');
        expect( pikachuTest.style.display).toBe('none')
        
    });

    
    test('debe mostrar un error si no se encuentra el hero (batman123)', () => { 
    
            render(
                <MemoryRouter initialEntries={['/search?q=batman123']}>
                    <SearchPage/>
                </MemoryRouter>
            );
           
            const  pikachuTest = screen.getByLabelText('pikachu');
            expect( pikachuTest.style.display ).toBe('');
    
    
    
    });

   
    
    test('debe llamar el navigate a la pantalla nueva', () => { 
        
        const inputValue = 'superman';
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input =  screen.getByRole('textbox');
        fireEvent.change( input, { target: {name: 'searchText', value: inputValue }});

        const form = screen.getByRole('form');
        fireEvent.submit( form );
        screen.debug();
        expect( mockUseNavigate ).toHaveBeenCalledWith(`?q=${ inputValue }`);

    });
        


});