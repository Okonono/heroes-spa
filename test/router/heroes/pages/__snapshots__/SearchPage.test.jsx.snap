// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Búsquedas
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="buscarTexto"
          placeholder="Buscar aquí"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-info mt-2"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Buscar un héroe
      </div>
      <div
        aria-label="pikachu"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hay resultados con 
        <b>
           
           
        </b>
      </div>
    </div>
  </div>
</div>
`;
